Replicator/Worker.hh

```diff
@@ -132,11 +133,19 @@ namespace litecore { namespace repl {
         virtual void afterEvent() override;
         virtual void caughtException(const std::exception& x) override;
 
-        // Add the token for collection info to the format string
+        // Add the token for collection info to the format string (and cache the string, returning the pointer)
+        // Concurrent read-write of cache is technically safe, but if a rehash occurs (capacity of _formatCache is
+        // exceeded), all iterators are invalidated. So we use a shared_mutex, with a shared_lock on reads and a
+        // unique_lock on writes - this allows multiple concurrent reads, but blocks reads when a write needs to occur
         static inline const char* formatWithCollection(const char* fmt) {
             const std::string fmtStr = format("%s %s", kCollectionLogFormat, fmt);
-            const auto        found  = _formatCache.insert({fmtStr, 0});
-            return found.first->first.data();
+            {
+                std::shared_lock sharedLock(_formatMutex);  // Multiple threads can read concurrently
+                const auto       found = _formatCache.find(fmtStr);
+                if ( found != _formatCache.end() ) { return found->data(); }
+            }
+            std::unique_lock lock(_formatMutex);  // Block all other threads if we need to perform an insert
+            return _formatCache.insert(fmtStr).first->data();
```

Comment:
I think we are over-engineering this. Why can't we just lock `_formatCache.insert(fmtStr).first->data();`. 

Note that based on the doc, insert() also check if the item exists or not.

Reply:
I assume you mean just basically
```c++
static inline const char* formatWithCollection(const char* fmt) {
        const std::string fmtStr = format("%s %s", kCollectionLogFormat, fmt);
        std::unique_lock lock(_formatMutex);
        const auto found = _formatCache.insert(fmtStr);
        return found.first->data();
    }
```
This method has the largest performance hit, as only a single thread can read or write at once. The implementation of the latest commit allows concurrent reads from any number of threads, until a write needs to occur.